routers/api/v1/repo/pull.go

```diff
@@ -1057,49 +1057,53 @@ func MergePullRequest(ctx *context.APIContext) {
 	}
 	log.Trace("Pull request merged: %d", pr.ID)
 
-	if form.DeleteBranchAfterMerge {
-		// Don't cleanup when there are other PR's that use this branch as head branch.
-		exist, err := issues_model.HasUnmergedPullRequestsByHeadInfo(ctx, pr.HeadRepoID, pr.HeadBranch)
-		if err != nil {
-			ctx.ServerError("HasUnmergedPullRequestsByHeadInfo", err)
-			return
-		}
-		if exist {
-			ctx.Status(http.StatusOK)
-			return
-		}
-
-		var headRepo *git.Repository
-		if ctx.Repo != nil && ctx.Repo.Repository != nil && ctx.Repo.Repository.ID == pr.HeadRepoID && ctx.Repo.GitRepo != nil {
-			headRepo = ctx.Repo.GitRepo
-		} else {
-			headRepo, err = gitrepo.OpenRepository(ctx, pr.HeadRepo)
+	if form.DeleteBranchAfterMerge && pr.Flow == issues_model.PullRequestFlowGithub {
+		// check permission even it has been checked in repo_service.DeleteBranch so that we don't need to
+		// do RetargetChildrenOnMerge
+		if err := repo_service.CanDeleteBranch(ctx, pr.HeadRepo, pr.HeadBranch, ctx.Doer); err == nil {
```

Comment:
I do not think the check is necessary or right.

![image](https://github.com/user-attachments/assets/aae092a4-9bcd-4525-baa7-67e46e09a009)

Reply:
`RetargetChildrenOnMerge`  already checks `pr.BaseRepoID == pr.HeadRepoID`, then in the same repo, the merger should also be able to delete?

Reply:
And TBH I don't quite understand the logic, so no idea whether it is right.

Reply:
> `RetargetChildrenOnMerge` already checks `pr.BaseRepoID == pr.HeadRepoID`, then in the same repo, the merger should also be able to delete?

I also noticed this logic and I don't think it's right, why only the same repository? But I will only fix the bug here and leave the refactor to other pull requests.